### DISCUSSION TITLE
how to use multiple graph in java client server mode?

### DISCUSSION THREAD

**Discussion by jinruin** [JanusGraph/janusgraph#2463] [Q&A]
As  our app must surpport multi tenants.
In embeded mode, we use ConfiguredGraphFactory to create new graph when new tenant registered.
But while multiple graph instance running, the index keep installed, can't transfer to enabled state.
So we are going to change to client-server mode. 
I don't konw how to connect multiple graph in client-server mode.
May someone hellp me?

**Comment by li-boxuan** [comment]
@BrunoBerisso Are you able to help here? I believe you are familiar with multi-tenants use case 😉

**Comment by jinruin** [comment]
yeah, I have some experiences in multi tenant appliacation.  by ur nick，i think ur chinese，right？ ------------------&nbsp;原始邮件&nbsp;------------------
发件人:&nbsp;&quot;Boxuan&amp;nbsp;Li&quot;<notifications@github.com&gt;
发送时间:&nbsp;2021年2月19日(星期五) 晚上10:34
收件人:&nbsp;&quot;JanusGraph/janusgraph&quot;<janusgraph@noreply.github.com&gt;;
抄送:&nbsp;&quot;jinruin&quot;<6969624@qq.com&gt;;&quot;Author&quot;<author@noreply.github.com&gt;;
主题:&nbsp;Re: [JanusGraph/janusgraph] how to use multiple graph in java client server mode? (#2463)

**Reply by li-boxuan** [reply]
This is a bit off-topic, but yes, I am :)

**Comment by BrunoBerisso** [comment]
When you open a graph with `ConfiguredGraphFactory` it will automatically add a traversal source to the Gremlin engine with the name "[your graph name]_traversal", check the [docs here ](https://docs.janusgraph.org/basics/configured-graph-factory/#graph-and-traversal-bindings) for more details. You can use that traversal source to query the graphs dynamically like this:

```java
import static org.apache.tinkerpop.gremlin.process.traversal.AnonymousTraversalSource.traversal;

String[] addresses = ["janus-1", "janus-2", "janus-3"];
Cluster cluster = Cluster.build().addContactPoints(addresses).create();

private GraphTraversalSource getGraphTraversalSource(String graphName) {
    if (graphName == null) {
        log.error("Bad graph name");
        return null;
    }
    return traversal().withRemote(DriverRemoteConnection.using(cluster, graphName + "_traversal"));
}
```

Keep in mind that a graph can take up to 20s to be available in all the nodes in the cluster.

**Reply by jinruin** [reply]
I create the graph 'g1' by gremlin client console.  And open it. 
Then I use the following code as your written:
``` java
        Cluster cluster = Cluster.build().addContactPoint("localhost").create();
        GraphTraversalSource g = traversal().withRemote(DriverRemoteConnection.using(cluster, "g1_traversal"));
        Long count = g.V().count().next();
        System.out.println("count:" + count);
```
But I got the error:
`
ResponseException: The traversal source [g1_traversal] for alias [g] is not configured on the server.
`

**Reply by jinruin** [reply]
I have tried many times as the doc.  
As the doc says: after ConfiguredGraphFactory.open("graph1") , you can use graph1 and graph1_traversal.
But I open the graph 'g1', I can use none of them. 
If i reconnect, i can use g1, but g1_traversal is still not available.
The graph version is 0.5.3

Following is the whole test output:
```
gremlin> :remote connect tinkerpop.server conf/remote.yaml session
14:47:13 WARN  org.apache.tinkerpop.gremlin.driver.Host  - Marking Host{address=localhost/127.0.0.1:8182, hostUri=ws://localhost:8182/gremlin} as unavailable. Trying to reconnect.
==>Configured localhost/127.0.0.1:8182-[694c54d2-c082-4536-ae58-3fd1f271c494]
gremlin> :remote console
==>All scripts will now be sent to Gremlin Server - [localhost/127.0.0.1:8182]-[694c54d2-c082-4536-ae58-3fd1f271c494] - type ':remote console' to return to local mode
gremlin> g1
No such property: g1 for class: Script2
Type ':help' or ':h' for help.
Display stack trace? [yN]
gremlin> ConfiguredGraphFactory.open('g1')
==>standardjanusgraph[hbase:[127.0.0.1]]
gremlin> g1
No such property: g1 for class: Script2
Type ':help' or ':h' for help.
Display stack trace? [yN]
gremlin> g1_traversal
No such property: g1_traversal for class: Script6
Type ':help' or ':h' for help.
Display stack trace? [yN]
gremlin> :remote console
==>All scripts will now be evaluated locally - type ':remote console' to return to remote mode for Gremlin Server - [localhost/127.0.0.1:8182]-[86cd2c88-76a6-42b4-9105-b0fb45ea14ae]
gremlin> :remote console
==>All scripts will now be sent to Gremlin Server - [localhost/127.0.0.1:8182]-[86cd2c88-76a6-42b4-9105-b0fb45ea14ae] - type ':remote console' to return to local mode
gremlin> g1
==>standardjanusgraph[hbase:[127.0.0.1]]
gremlin> g1_traversal
No such property: g1_traversal for class: Script8
Type ':help' or ':h' for help.
Display stack trace? [yN]
```

**Reply by BrunoBerisso** [reply]
Not sure how are you creating the graphs but I think the bindings in the Gremlin engine will not be updated for already opened connections. Try reconnecting instead of changing the `console` mode, I mean:

1. Open a gremlin shell and create the graph
2. Close the shell
3. Open a new gremlin shell and try to access the bindings `g1` and `g1_traversal`

**Reply by melgenek** [reply]
If you're using the docker for janusgraph, then you have the tinkerpop's channelizer by default
```
channelizer: org.apache.tinkerpop.gremlin.server.channel.WebSocketChannelizer
```
Make sure that you replacing it with the Janus' channelizer before connecting remotely:
```
channelizer: org.janusgraph.channelizers.JanusGraphWebSocketChannelizer
```

**Reply by farodin91** [reply]
Just as a glimpse, I work on solution to get rid of the JanusGraph special channelizer. https://github.com/JanusGraph/janusgraph/pull/2492

**Reply by melgenek** [reply]
@farodin91 that is amazing :smile: I wish this feature was released before the previous week when I was racking my brain to solve this issue.
